core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -160,55 +172,74 @@ public void clusterChanged(final ClusterChangedEvent event) {
-            applyCleanedIndices(event);
-            // make sure that newly created shards use the latest meta data
-            applyIndexMetaData(event);
-            applyNewIndices(event);
-            // apply mappings also updates new indices. TODO: make new indices good to begin with
-            applyMappings(event);
-            applyNewOrUpdatedShards(event);
+            updateFailedShardsCache(event);
+
+            deleteIndices(event); // also deletes shards of deleted indices
+
+            removeIndices(event); // also removes shards of removed indices
+
+            removeShards(event);
+
+            updateIndices(event); // can also fail shards, but these are then guaranteed to be in failedShardsCache
+
+            createIndices(event);
+
+            createOrUpdateShards(event);
         }
     }
 
-    private void cleanFailedShards(final ClusterChangedEvent event) {
-        RoutingNode routingNode = event.state().getRoutingNodes().node(event.state().nodes().getLocalNodeId());
-        if (routingNode == null) {
-            failedShards.clear();
+    /**
+     * Removes shard entries from the failed shards cache that are no longer allocated to this node by the master.
+     * Sends shard failures for shards that are marked as actively allocated to this node but don't actually exist on the node.
+     * Resends shard failures for shards that are still marked as allocated to this node but previously failed.
+     *
+     * @param event cluster change event
+     */
+    private void updateFailedShardsCache(final ClusterChangedEvent event) {
+        RoutingNode localRoutingNode = event.state().getRoutingNodes().node(event.state().nodes().getLocalNodeId());
+        if (localRoutingNode == null) {
+            failedShardsCache.clear();
             return;
         }
-        for (Iterator<Map.Entry<ShardId, ShardRouting>> iterator = failedShards.entrySet().iterator(); iterator.hasNext(); ) {
-            Map.Entry<ShardId, ShardRouting> entry = iterator.next();
-            ShardRouting failedShardRouting = entry.getValue();
-            ShardRouting matchedShardRouting = routingNode.getByShardId(failedShardRouting.shardId());
+
+        // remove items from cache which are not in our routing table anymore
+        for (Iterator<Map.Entry<ShardId, ShardRouting>> iterator = failedShardsCache.entrySet().iterator(); iterator.hasNext(); ) {
+            ShardRouting failedShardRouting = iterator.next().getValue();
+            ShardRouting matchedShardRouting = localRoutingNode.getByShardId(failedShardRouting.shardId());
             if (matchedShardRouting == null || matchedShardRouting.isSameAllocation(failedShardRouting) == false) {
                 iterator.remove();
             }
         }
+
+        DiscoveryNode masterNode = event.state().nodes().getMasterNode();
+        for (ShardRouting shardRouting : localRoutingNode) {
+            if (failedShardsCache.containsKey(shardRouting.shardId())) {
```

Comment:
I wonder if we can move this part to the previous iteration as an else of `if matchedShardRouting==null` , which will allow us to take this loop and put it in it's own method called "failMissingShards".. wdyt?